README.md
# Convolutional-Neural-Network

## 1 DeepCrack
This a convolutional Neural Network just as U-Net to segment pavement crack from background. This code is taken from source of DeepCrack (https://github.com/qinnzou/DeepCrack). We are planning to fine tune DeepCrack 
on our pavement distress datasets. We are going through several test runs on CRACKTRE260 and EdmCrack600 and other published pavement distress datasets. 

Before using this model to your research cite from Zou Q, Zhang Z, Li Q, Qi X, Wang Q and Wang S, DeepCrack: Learning Hierarchical Convolutional Features for Crack Detection, IEEE Transactions on Image Processing, vol. 28, no. 3, pp. 1498-1512, 2019.

## 2 VGG 19 

Applied VGG 19 to predict the class of cracks in EdmCrack 600. I categorizied all of the 600 cracks into Longitudinal, Alligator, Transverse and Delaminations . Then tuned the VGG19 with different hyperparameters to predict the pavemenent distress types. 

Ref: Q. Mei, M. Gül, and M.R. Azim, Densely connected deep neural network considering connectivity of pixels for automatic crack detection. Automation in Construction, 2020. 110: p. 103018. 


DeepCrack/history.txt
   18  pwd
   19  pwd 
   20   ls
   21  cd data
   22  ls
   23  vi train_example.txt 
   24  ls
   25  cd DeepCrack_Master/
   26  ls
   27  cd codes/
   28  ls
   29  shbatch shellscript.sh 
   30  sbatch shellscript.sh 
   31  squeue -u tadnan
   32  sbatch shellscript.sh 
   33  squeue -u tadnan
   34  sbatch shellscript.sh 
   35  squeue -u tadnan
   36  sbatch shellscript.sh 
   37  squeue -u tadnan
   38  sbatch shellscript.sh 
   39  squeue -u tadnan
   40  sbatch shellscript.sh 
   41  ls
   42  cd DeepCrack_Master/
   43  ls
   44  cd codes
   45  ls
   46  sbatch shellscript.sh 
   47  squeue -u tadnan
   48  sbatch shellscript.sh 
   49  squeue -u tadnan
   50  sbatch shellscript.sh 
   51  squeue -u tadnan
   52  sbatch shellscript.sh 
   53  squeue -u tadnan
   54  sbatch shellscript.sh 
   55  squeue -u tadnan
   56  sbatch shellscript.sh 
   57  squeue -u tadnan
   58  sbatch shellscript.sh 
   59  squeue -u tadnan
   60  sbatch shellscript.sh 
   61  squeue -u tadnan
   62  sbatch shellscript.sh 
   63  squeue -u tadnan
   64  sbatch shellscript.sh 
   65  squeue -u tadnan
   66  sbatch shellscript.sh 
   67  ls
   68  cd DeepCrack_Master/
   69  cd codes/
   70  ls
   71  sbatch shellscript.sh 
   72  ls
   73  cd DeepCrack_Master/
   74  cd cdoes
   75  cd codes
   76  ls
   77  sbatch shellscript.sh 
   78  squeue -u tadnan
   79  sbatch shellscript.sh 
   80  squeue -u tadnan
   81  sbatch shellscript.sh 
   82  squeue -u tadnan
   83  ls
   84  cd DeepCrack_Master/
   85  ls
   86  cd codes/
   87  ls
   88  sbatch shellscript.sh 
   89  squeue -u tadnan
   90  sbatch shellscript.sh 
   91  squeue -u tadnan
   92  sbatch shellscript.sh 
   93  squeue -u tadnan
   94  sbatch shellscript.sh 
   95  squeue -u tadnan
   96  ls
   97  pwd
   98  sbatch shellscript.sh 
   99  squeue -u tadnan
  100  cd ..
  101  ls
  102  cd 'DeepCrack_Master (with Zach&Tianyang)'/
  103  ls
  104  cd codes
  105  ls
  106  sbatch shellscript.sh 
  107  squeue -u tadnan
  108  cd ..
  109  pwd
  110  ls
  111  cd DeepCrack_Master
  112  cd codes/
  113  ls
  114  sbatch shellscript.sh 
  115  squeue -u tadnan
  116  sbatch shellscript.sh 
  117  squeue -u tadnan
  118  sbatch shellscript.sh 
  119  squeue -u tadnan
  120  ls 
  121  cd DeepCrack_Master
  122  cd codes/
  123  ls
  124  sbatch shellscript.sh 
  125  squeue -u tadnan
  126  ls
  127  cd DeepCrack_Master
  128  ls
  129  cd codes/
  130  ls
  131  sbatch shellscript.sh 
  132  squeue -u tadnan
  133  cd ..
  134  ls
  135  cd ..
  136  ls
  137  cd EdmCrack600_DeepCrack/
  138  ls
  139  cd codes/
  140  ls
  141  sbatch shellscript.sh 
  142  squeue -u tadnan
  143  sbatch shellscript.sh 
  144  squeue -u tadnan
  145  sbatch shellscript.sh 
  146  squeue -u tadnan
  147  sbatch shellscript.sh 
  148  squeue -u tadnan
  149  sbatch test.sh
  150  squeue -u tadnan
  151  sbatch test.sh
  152  squeue -u tadnan
  153  sbatch shellscript.sh 
  154  squeue -u tadnan
  155  sbatch test.sh
  156  sbatch shellscript.sh 
  157  squeue -u tadnan
  158  sbatch shellscript.sh 
  159  squeue -u tadnan
  160  sbatch shellscript.sh 
  161  squeue -u tadnan
  162  sbatch shellscript.sh 
  163  squeue -u tadnan
  164  scancel 771581
  165  squeue -u tadnan
  166  sbatch shellscript.sh 
  167  squeue -u tadnan
  168  cd ..
  169   cd..
  170  ls
  171  cd ..
  172  ls
  173  cd ..
  174  ls
  175  cd tadnan
  176  ls
  177  cd DeepCrack_Master
  178  ls
  179  cd codes/
  180  ls
  181  sbatch shellscript.sh 
  182  squeue -u tadnan
  183  cd ..
  184  ls
  185  cd EdmCrack600_DeepCrack/
  186  cd codes
  187  ls
  188  sbatch test.sh 
  189  squeue -u tadnan
  190  ls
  191  pwd
  192  sbatch shellscript.sh 
  193  cd ..
  194  ls
  195  cd DeepCrack_Master
  196  ls
  197  cd codes
  198  ls
  199  sbtach test.sh
  200  sbatch test.sh
  201  squeue -u tadnan
  202  pwd
  203  cd ..
  204  ls
  205  cd EdmCrack600_DeepCrack/
  206  ls
  207  cd codes/
  208  ls
  209  sbtach test.sh
  210  pwd
  211  sbatch test.sh
  212  cd DeepCrack_Master
  213  cd codes/
  214  sbatch shellscript.sh 
  215  squeue -u tadnan
  216  scancel 871941
  217  squeue -u tadnan
  218  ls
  219  cd DeepCrack_Master
  220  pwd
  221  cd codes/
  222  ls
  223  sbatch shellscript.sh 
  224  squeue -u tadnan
  225  scancel 872592
  226  sbatch shellscript.sh 
  227  squeue -u tadnan
  228  scancel 875002
  229  squeue -u tadnan
  230  sbatch shellscript.sh 
  231  squeue -u tadnan
  232  scancel 875003
  233  sbatch shellscript.sh 
  234  squeue -u tadnan
  235  scancel 875006
  236  scancel 875007
  237  squeue -u tadnan
  238  sbatch shellscript.sh 
  239  squeue -u tadnan
  240  scancel 875009
  241  squeue -u tadnan
  242  sbatch shellscript.sh 
  243  squeue -u tadnan
  244  sbatch shellscript.sh 
  245  squeue -u tadnan
  246  sbatch shellscript.sh 
  247  squeue -u tadnan
  248  sbatch shellscript.sh 
  249  squeue -u tadnan
  250  sbatch shellscript.sh 
  251  squeue -u tadnan
  252  scancel 879924
  253  sbatch shellscript.sh 
  254  squeue -u tadnan
  255  sbatch shellscript.sh 
  256  squeue -u tadnan
  257  sbatch shellscript.sh 
  258  squeue -u tadnan
  259  sbatch shellscript.sh 
  260  squeue -u tadnan
  261  ls
  262  sbatch shellscript.sh 
  263  cd DeepCrack_Master
  264  cd codes/
  265  sbatch shellscript.sh 
  266  squeue -u tadnan
  267  scancel 884324
  268  sbatch shellscript.sh 
  269  squeue -u tadnan
  270  sbatch shellscript.sh 
  271  squeue -u tadnan
  272  sbatch shellscript.sh 
  273  squeue -u tadnan
  274  scanel 884339
  275  scancel 884339
  276  sbatch shellscript.sh 
  277  squeue -u tadnan
  278  scancel 884339
  279  scancel 884341
  280  squeue -u tadnan
  281  sbatch shellscript.sh 
  282  squeue -u tadnan
  283  sbatch shellscript.sh 
  284  squeue -u tadnan
  285  sbatch shellscript.sh 
  286  squeue -u tadnan
  287  sbatch shellscript.sh 
  288  squeue -u tadnan
  289  ls
  290  cd DeepCrack_Master
  291   cd codes/

UNets/Encoder_decoder_on EdmCrack600.py
#!/usr/bin/env python
# coding: utf-8

# In[70]:


#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# License: BSD
# Author: Sasank Chilamkurthy

from __future__ import print_function, division

import torch
import torch.nn as nn
import torch.optim as optim
from torch.optim import lr_scheduler
import torch.backends.cudnn as cudnn
import numpy as np
import torchvision
from torchvision import datasets, models, transforms
from torch.utils.data import random_split
import matplotlib.pyplot as plt
import time
import os
import copy

cudnn.benchmark = True
plt.ion()   # interactive mode

from random import *
from tqdm.notebook import tqdm, trange
from time import sleep
from pathlib import Path
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import transforms, models
from torchvision.datasets import ImageFolder
from warnings import filterwarnings
filterwarnings('ignore')

  
# functions to show an image

def imshow(img):
    img = img / 2 + 0.5     # unnormalize
    npimg = img.numpy()
    plt.imshow(np.transpose(npimg, (1, 2, 0)))
    plt.show()
    plt.savefig('labels.JPG')
    
##  codes for data augmentation
train_trans = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.RandomHorizontalFlip(p=0.5), ## tamim: image will move left and right
    transforms.RandomVerticalFlip(p=0.5), ## tamim: image will come to eye vertically
    transforms.RandomRotation(degrees=(.5, 5)), ## very small rotation of the cracks
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.5, 0.5, 0.5],
        std=[0.5, 0.5, 0.5]
    )
])


test_trans = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.RandomHorizontalFlip(p=0.5), ## tamim: image will move left and right
    transforms.RandomVerticalFlip(p=0.5), ## tamim: image will come to eye vertically
    transforms.RandomRotation(degrees=(.5, 5)), ## very small rotation of the cracks
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.5, 0.5, 0.5],
        std=[0.5, 0.5, 0.5]
    )
])

##  codes for data augmentation
train_trans = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.RandomHorizontalFlip(p=0.5), ## tamim: image will move left and right
    transforms.RandomVerticalFlip(p=0.5), ## tamim: image will come to eye vertically
    transforms.RandomRotation(degrees=(.5, 5)), ## very small rotation of the cracks
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.5, 0.5, 0.5],
        std=[0.5, 0.5, 0.5]
    )
])


test_trans = transforms.Compose([
    transforms.Resize((224, 224)),
    transforms.RandomHorizontalFlip(p=0.5), ## tamim: image will move left and right
    transforms.RandomVerticalFlip(p=0.5), ## tamim: image will come to eye vertically
    transforms.RandomRotation(degrees=(.5, 5)), ## very small rotation of the cracks
    transforms.ToTensor(),
    transforms.Normalize(
        mean=[0.5, 0.5, 0.5],
        std=[0.5, 0.5, 0.5]
    )
])

## Load data
from torchvision.datasets import ImageFolder

data = ImageFolder("D:\\SPRING 2023\\MS_Thesis\\MASTER'S THESIS\\Thesis results\\Data\\Data_Structure(Annotated)", transform=train_trans , )

test_folder= ImageFolder("D:\\SPRING 2023\\MS_Thesis\\MASTER'S THESIS\\DATA_iPhone_13_Pro_Max", transform=test_trans, )

batch_size = 32
num_classes = 4
learning_rate = 0.0001
num_epochs = 10

print("Follwing classes are there : \n",data.classes)

classes = ('Alligator Cracks', 'Delamination', 'Longitudinal Cracks', 'Transverse Cracks')

len(data)

##Splitting Data and Prepare Batches: 
## Source: https://medium.com/thecyphy/train-cnn-model-with-pytorch-21dafb918f48 
val_size = 127 ##  Tamim:30% data for validation ## 
train_size = len(data) - val_size 


train_loader,val_loader = random_split(data,[train_size,val_size]) ## To randomly split the images into training and testing, PyTorch provides random_split()
print(f"Length of Train Data : {len(train_loader)}")                   ## changed the folder names
print(f"Length of Validation Data : {len(val_loader)}")

# Splitting train and validation data on batches

train_loader = torch.utils.data.DataLoader(train_loader, shuffle=True, batch_size=batch_size) ## defined train data & val data
val_loader = torch.utils.data.DataLoader(val_loader, shuffle=True, batch_size=batch_size)
test_loader = torch.utils.data.DataLoader(test_folder, shuffle=False, batch_size=batch_size)


# visualize images of a single batch

dataiter = iter(train_loader)
images, labels = next(dataiter)

# show images
imshow(torchvision.utils.make_grid(images))
# print labels
print(' '.join(f'{classes[labels[j]]:5s}' for j in range(batch_size)))


## define model

class Block(nn.Module):
    def __init__(self, in_ch, out_ch):
        super().__init__()
        self.conv1 = nn.Conv2d(in_ch, out_ch, 3)
        self.relu  = nn.ReLU()
        self.conv2 = nn.Conv2d(out_ch, out_ch, 3)
    
    def forward(self, x):
        return self.conv2(self.relu(self.conv1(x)))


class Encoder(nn.Module):
    def __init__(self, chs=(3,64,128,256,512,1024)):
        super().__init__()
        self.enc_blocks = nn.ModuleList([Block(chs[i], chs[i+1]) for i in range(len(chs)-1)])
        self.pool       = nn.MaxPool2d(2)
    
    def forward(self, x):
        ftrs = []
        for block in self.enc_blocks:
            x = block(x)
            ftrs.append(x)
            x = self.pool(x)
        return ftrs


class Decoder(nn.Module):
    def __init__(self, chs=(1024, 512, 256, 128, 64)):
        super().__init__()
        self.chs         = chs
        self.upconvs    = nn.ModuleList([nn.ConvTranspose2d(chs[i], chs[i+1], 2, 2) for i in range(len(chs)-1)])
        self.dec_blocks = nn.ModuleList([Block(chs[i], chs[i+1]) for i in range(len(chs)-1)]) 
        
    def forward(self, x, encoder_features):
        for i in range(len(self.chs)-1):
            x        = self.upconvs[i](x)
            enc_ftrs = self.crop(encoder_features[i], x)
            x        = torch.cat([x, enc_ftrs], dim=1)
            x        = self.dec_blocks[i](x)
        return x
    
    def crop(self, enc_ftrs, x):
        _, _, H, W = x.shape
        enc_ftrs   = torchvision.transforms.CenterCrop([H, W])(enc_ftrs)
        return enc_ftrs

    
## Change the class 4 from 1

class UNet(nn.Module):
    def __init__(self, enc_chs=(3,64,128,256,512,1024), dec_chs=(1024, 512, 256, 128, 64), num_class=4, retain_dim=False, out_sz=(572,572)):
        super().__init__()
        self.encoder     = Encoder(enc_chs)
        self.decoder     = Decoder(dec_chs)
        self.head        = nn.Conv2d(dec_chs[-1], num_class, 1)
        self.retain_dim  = retain_dim

    def forward(self, x):
        enc_ftrs = self.encoder(x)
        out      = self.decoder(enc_ftrs[::-1][0], enc_ftrs[::-1][1:])
        out      = self.head(out)
        if self.retain_dim:
            out = F.interpolate(out, out_sz)
        return out
    
#https://amaarora.github.io/2020/09/13/unet.html
#https://medium.com/analytics-vidhya/unet-implementation-in-pytorch-idiot-developer-da40d955f201

model = UNet()

# specify loss function
criterion = nn.CrossEntropyLoss()

# specify loss function
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# loop over epochs
print("[INFO] training the network...")
startTime = time.time()
for e in tqdm(range(num_epochs)):
	# set the model in training mode
	model.train()
	# initialize the total training and validation loss
	totalTrainLoss = 0
	totalTestLoss = 0
	# loop over the training set
	for (i, (x, y)) in enumerate(train_loader):
		# send the input to the device
		#(x, y) = (x.to(config.DEVICE), y.to(config.DEVICE))
		# perform a forward pass and calculate the training loss
		pred = model(x)
		loss = criterion(pred, y)
		# first, zero out any previously accumulated gradients, then
		# perform backpropagation, and then update model parameters
		opt.zero_grad()
		loss.backward()
		opt.step()
		# add the loss to the total training loss so far
		totalTrainLoss += loss
	# switch off autograd
	with torch.no_grad():
		# set the model in evaluation mode
		model.eval()


# In[ ]:




